domain/modelmigration/service/service.go

```diff
@@ -0,0 +1,163 @@
+	// [ResourceProvider]
+	resourceProviderGettter func(context.Context) (ResourceProvider, error)
+
+	st State
+}
+
+// State defines the interface required for accessing the underlying state of
+// the model during migration.
+type State interface {
+	GetControllerUUID(context.Context) (string, error)
+}
+
+// New is responsible for constructing a new [Service] to handle model migration
+// tasks.
+func New(
+	instanceProviderGetter providertracker.ProviderGetter[InstanceProvider],
+	resourceProviderGetter providertracker.ProviderGetter[ResourceProvider],
+	st State,
+) *Service {
+	return &Service{
+		instanceProviderGetter:  instanceProviderGetter,
+		resourceProviderGettter: resourceProviderGetter,
+		st:                      st,
+	}
+}
+
+// AdoptResources is responsible for taking ownership of the cloud resources of
+// a model when it has been migrated into this controller.
+func (s *Service) AdoptResources(
+	ctx context.Context,
+	sourceControllerVersion version.Number,
+) error {
+	provider, err := s.resourceProviderGettter(ctx)
+
+	// Provider doesn't support adopting resources and this is ok!
+	if errors.Is(err, coreerrors.NotSupported) {
+		return nil
+	} else if err != nil {
+		return errors.Errorf(
+			"getting resource provider for adopting model cloud resources: %w",
+			err,
+		)
+	}
+
+	controllerUUID, err := s.st.GetControllerUUID(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"cannot get controller uuid while adopting model cloud resources: %w",
+			err,
+		)
+	}
+
+	err = provider.AdoptResources(
+		environscontext.WithoutCredentialInvalidator(ctx),
+		controllerUUID,
+		sourceControllerVersion,
+	)
+
+	// Provider doesn't support adopting resources and this is ok!
+	if errors.Is(err, coreerrors.NotImplemented) {
+		return nil
+	}
+	if err != nil {
+		return errors.Errorf("cannot adopt cloud resources for model: %w", err)
+	}
+	return nil
+}
+
+// CheckMachines is responsible for checking a model after it has been migrated
+// into this target controller. We check the machines that exist in the model
+// against the machines reported by the models cloud and report any
+// discrepancies.
+func (s *Service) CheckMachines(
+	ctx context.Context,
+) ([]modelmigration.MigrationMachineDiscrepancy, error) {
+	provider, err := s.instanceProviderGetter(ctx)
+	if err != nil && !errors.Is(err, coreerrors.NotSupported) {
+		return nil, errors.Errorf(
+			"cannot get provider for model to check machines provider machines against the controller: %w",
+			err,
+		)
+	}
+
+	// If the provider doesn't support machines we can bail out early.
+	if errors.Is(err, coreerrors.NotSupported) {
+		return nil, nil
+	}
+
+	_, err = provider.AllInstances(envcontext.WithoutCredentialInvalidator(ctx))
+	if err != nil {
+		return nil, errors.Errorf(
+			"cannot get all provider instances for model when checking machines: %w",
+			err,
+		)
+	}
+
+	// TODO (tlm) 28/7/2024: This function is incomplete at the moment till we
```

Comment:
Prefer a TODO reflecting the domain area, eg `TODO(instancedata)` or `TODO(machines)`
That way all the TODOs for a given epic can be easily searched for

Reply:
Done & Done